core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -2496,8 +2496,23 @@ object Stream {
      * For example, `merge` is implemented by calling `unconsAsync` on each stream, racing the
      * resultant `AsyncPull`s, emitting winner of the race, and then repeating.
      */
-    def unconsAsync(implicit ec: ExecutionContext, F: Effect[F]): Pull[F,Nothing,AsyncPull[F,Option[(Segment[O,Unit], Stream[F,O])]]] =
-      Pull.fromFreeC(Algebra.unconsAsync(self.get, ec)).map(_.map(_.map { case (hd, tl) => (hd, Stream.fromFreeC(tl)) }))
+    def unconsAsync(implicit ec: ExecutionContext, F: Effect[F]): Pull[F,Nothing, AsyncPull[F,Option[(Segment[O,Unit], Stream[F,O])]]] = {
+      type UO = Option[(Segment[O,Unit], FreeC[Algebra[F,O,?],Unit])]
+
+      Pull.fromFreeC {
+        val ref = new async.Ref[F, Either[Throwable, Option[(Segment[O,Unit], Stream[F,O])]]]
+        Algebra.getScope[F, Nothing] flatMap { scope =>
+          val runStep =
+            Algebra.runFoldScope(
+              scope
+              , Algebra.uncons(self.get).flatMap(Algebra.output1(_))
+              , None : UO
+            ){ (_, uo) => uo.asInstanceOf[UO] } map { _ map { case (hd, tl) => (hd, fromFreeC(tl)) }}
+
+          Algebra.eval(ref.setAsync(F.attempt(runStep))) map { _ => AsyncPull.readAttemptRef(ref) }
```

Comment:
As per the current implementation of `Ref`, `attempt(runStep)` is redundant, and `readAttemptRef` can take a `Ref[F, A]` rather than a `Ref[Throwable, A]` (with some implementation changes). However, I much prefer to be explicit just like you were: I think we should remove the internal throwable from Ref, and use a `Ref[F, Either[Throwable, A]` explicitly when we want to carry an exception within it.

Reply:
I think `setAsync` has weird semantics. Why does it fork the `F[A]` evaluation and why can't the caller do that instead? If the caller did the fork, then we'd be able to implement `setAsync` in a way that doesn't fail.

Reply:
I was sort of puzzled by this as well. I guess we have to trust that there were enough discussions on the Cats.Effects for this. Didn't follow all that recently, so I really don't have idea why that changed from the way how we had implemented it in 0.9.

Reply:
@SystemFw lets revisit Ref implementation later. I think perhaps we can do a better with Ref. Also perhaps we can work with folks in CatsEffects to bring something that will be stronger.

Reply:
Unrelated to cats-effect stuff. You can see the old 0.9 stuff worked the same way: https://github.com/functional-streams-for-scala/fs2/blob/108f94a0a4a4e21bcc99a900db28e1ca54439570/core/shared/src/main/scala/fs2/Task.scala#L342-L343

Reply:
Ah @mpilquist I see what do you mean. Perhaps we shall give it a try ?

Reply:
hm, cannot we have all the set's in fact synchronous? I am curios if we need in fact actor for defining the Ref?

Reply:
Sure, I'm all for it assuming it doesn't break too much stuff. @SystemFw do you want to give this a shot?

Reply:
I've been thinking a lot about `Ref`, and I think it can be improved (or at least it's worth a try), in a separate PR (I also want to tackle #888 at some point, with a new Actor message).

But yeah, it's a separate discussion. Btw, I second the awesomeness of removing an essential (I thought) algebra primitive